caikit_nlp/modules/text_embedding/embedding.py

```diff
@@ -257,26 +305,46 @@ def _truncate_input_tokens(
             max_length = max_tokens
             ret = texts  # will not alter texts when truncation is not allowed
 
-        for text in texts:
-            tokenized = self.model.tokenizer(
-                text,
-                return_attention_mask=False,
-                return_token_type_ids=False,
-                return_overflowing_tokens=True,
-                return_length=True,
-                truncation=True,
-                max_length=max_length,
-            )
+        tokenized = self._with_retry(
+            self._tokenizer,
+            texts,
+            return_attention_mask=False,
+            return_token_type_ids=False,
+            return_overflowing_tokens=True,
+            return_offsets_mapping=True,
+            return_length=True,
+            truncation=True,
+            max_length=max_length,
+        )
 
-            lengths = tokenized["length"]
+        for i, text in enumerate(texts):
+            mapping = tokenized["overflow_to_sample_mapping"]
+            lengths = [
+                tokenized["length"][idx] for idx, v in enumerate(mapping) if v == i
+            ]
             was_truncated = len(lengths) > 1  # multiple lengths when truncated
 
             if okay_to_truncate and was_truncated:
-                # decode the truncated input tokens back to text to be returned
+                # Get the text offsets for the tokens that are kept after truncation
+                offsets = [
+                    tokenized["offset_mapping"][idx]
+                    for idx, v in enumerate(mapping)
+                    if v == i
+                ]
+                truncated_offsets = offsets[0]
```

Comment:
I refactored and commented.  It's still confusing but I think it's quite a bit simpler.

* The sample mapping is important because we run all the texts in tokenize() so basically that is a mask that helps us get the lengths and offsets for the one text we are handling (in the loop).

* lengths gives us the length of the truncated part and overflow parts if any.  So that tells us if it was truncated and helps report the number of tokens in the error message (when truncation not allowed -- default)

* offsets tell us what characters the tokens map to.  So we use that to extract the text to keep to fit this token limit  (instead of decoding which doesn't give us exactly what we passed in).  And to do that we just need to find the start of the first one and the end of the last one.   There might be a param so I could remove the check for (0, 0) but I think it's safer to leave it in for now.